src/shims/sync.rs

```diff
@@ -434,16 +532,34 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    // FIXME: We should check that this lock was locked by the active thread.
     fn pthread_rwlock_unlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
         let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
         let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        let writer_blockset = rwlock_get_or_create_writer_blockset(this, rwlock_op)?;
         if let Some(new_readers) = readers.checked_sub(1) {
+            assert_eq!(writers, 0);
             rwlock_set_readers(this, rwlock_op, Scalar::from_u32(new_readers))?;
+            if new_readers == 0 {
+                if let Some(_writer) = this.unblock_some_thread(writer_blockset)? {
+                    rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
+                }
+            }
             Ok(0)
         } else if writers != 0 {
+            let reader_blockset = rwlock_get_or_create_reader_blockset(this, rwlock_op)?;
             rwlock_set_writers(this, rwlock_op, Scalar::from_u32(0))?;
+            if let Some(_writer) = this.unblock_some_thread(writer_blockset)? {
+                rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
```

Comment:
Could you do this more symmetrically with `mutex`, and avoid calling `rwlock_set_writers` twice?

Reply:
Also there is a scheduling policy embedded here: writers get priority. Seems an okay choice, but please point out in a comment that we are making a policy decision here.